File: src/main/java/com/github/ccguyka/showupdates/consumer/ProjectUpdatesMerger.java
package com.github.ccguyka.showupdates.consumer;

import static java.util.stream.Collectors.toList;

import java.util.Set;
import com.github.ccguyka.showupdates.objects.ArtifactUpdate;
import com.github.ccguyka.showupdates.objects.DependencyUpdates;
import com.github.ccguyka.showupdates.objects.ProjectUpdates;
import com.google.common.collect.Sets;

class ProjectUpdatesMerger {

    public ProjectUpdates merge(ProjectUpdates first, ProjectUpdates second) {
        Set<ArtifactUpdate> parent = Sets.newHashSet();
        parent.addAll(first.getParent().getArtifacts());
        parent.addAll(second.getParent().getArtifacts());

        Set<ArtifactUpdate> dependency = Sets.newHashSet();
        dependency.addAll(first.getDependency().getArtifacts());
        dependency.addAll(second.getDependency().getArtifacts());

        Set<ArtifactUpdate> dependencyManagement = Sets.newHashSet();
        dependencyManagement.addAll(first.getDependencyManagement().getArtifacts());
        dependencyManagement.addAll(second.getDependencyManagement().getArtifacts());

        Set<ArtifactUpdate> plugin = Sets.newHashSet();
        plugin.addAll(first.getPlugin().getArtifacts());
        plugin.addAll(second.getPlugin().getArtifacts());

        return ProjectUpdates.builder()
                .withParent(new DependencyUpdates(parent.stream().collect(toList())))
                .withDependency(new DependencyUpdates(dependency.stream().collect(toList())))
                .withDependencyManagement(new DependencyUpdates(dependencyManagement.stream().collect(toList())))
                .withPlugin(new DependencyUpdates(plugin.stream().collect(toList())))
                .build();
    }
}

File: src/main/kotlin/com/github/ccguyka/showupdates/consumer/ProjectUpdatesMerger.kt
package com.github.ccguyka.showupdates.consumer

import com.github.ccguyka.showupdates.objects.ArtifactUpdate
import com.github.ccguyka.showupdates.objects.DependencyUpdates
import com.github.ccguyka.showupdates.objects.ProjectUpdates
import com.github.ccguyka.showupdates.objects.ProjectUpdates.Companion.builder
import com.google.common.collect.Sets
import java.util.stream.Collectors

internal class ProjectUpdatesMerger {

    fun merge(first: ProjectUpdates, second: ProjectUpdates): ProjectUpdates {
        val parent: MutableSet<ArtifactUpdate> = Sets.newHashSet()
        parent.addAll(first.parent.artifacts)
        parent.addAll(second.parent.artifacts)
        val dependency: MutableSet<ArtifactUpdate> = Sets.newHashSet()
        dependency.addAll(first.dependency.artifacts)
        dependency.addAll(second.dependency.artifacts)
        val dependencyManagement: MutableSet<ArtifactUpdate> = Sets.newHashSet()
        dependencyManagement.addAll(first.dependencyManagement.artifacts)
        dependencyManagement.addAll(second.dependencyManagement.artifacts)
        val plugin: MutableSet<ArtifactUpdate> = Sets.newHashSet()
        plugin.addAll(first.plugin.artifacts)
        plugin.addAll(second.plugin.artifacts)
        return builder()
                .withParent(DependencyUpdates(parent.stream().collect(Collectors.toList())))
                .withDependency(DependencyUpdates(dependency.stream().collect(Collectors.toList())))
                .withDependencyManagement(DependencyUpdates(dependencyManagement.stream().collect(Collectors.toList())))
                .withPlugin(DependencyUpdates(plugin.stream().collect(Collectors.toList())))
                .build()
    }
}
